lib/Analysis/IntervalAnalysis.cpp

```diff
@@ -93,17 +97,52 @@ mod(llvm::SMTSolverRef solver, const ExpressionValue &lhs, const ExpressionValue
   return res;
 }
 
+ExpressionValue
+bitAnd(llvm::SMTSolverRef solver, const ExpressionValue &lhs, const ExpressionValue &rhs) {
+  ExpressionValue res;
+  res.i = lhs.i & rhs.i;
+  res.expr = solver->mkBVAnd(lhs.expr, rhs.expr);
+  return res;
+}
+
+ExpressionValue
+shiftLeft(llvm::SMTSolverRef solver, const ExpressionValue &lhs, const ExpressionValue &rhs) {
+  ExpressionValue res;
+  res.i = lhs.i << rhs.i;
+  res.expr = solver->mkBVShl(lhs.expr, rhs.expr);
+  return res;
+}
+
+ExpressionValue
+shiftRight(llvm::SMTSolverRef solver, const ExpressionValue &lhs, const ExpressionValue &rhs) {
+  ExpressionValue res;
+  res.i = lhs.i >> rhs.i;
+  res.expr = solver->mkBVLshr(lhs.expr, rhs.expr);
+  return res;
+}
+
 ExpressionValue
 cmp(llvm::SMTSolverRef solver, CmpOp op, const ExpressionValue &lhs, const ExpressionValue &rhs) {
   ExpressionValue res;
-  res.i = Interval::Boolean(lhs.getField());
+  const Field &f = lhs.getField();
+  // Default result is any boolean output for when we are unsure about the comparison result.
+  res.i = Interval::Boolean(f);
   switch (op.getPredicate()) {
   case FeltCmpPredicate::EQ:
     res.expr = solver->mkEqual(lhs.expr, rhs.expr);
-    res.i = lhs.i.intersect(rhs.i);
+    if (lhs.i.isDegenerate() && rhs.i.isDegenerate() && lhs.i == rhs.i) {
+      res.i = Interval::True(f);
```

Comment:
Should both sides degenerate but `!=` give a `False` result here. Similar for the `NE` case below.